src/main/java/carpet/logging/TypeLogger.java

```diff
@@ -0,0 +1,141 @@
+            throw new NullPointerException("Value was not seen as valid: "+option);
+        }
+        parsedPlayerOptions.put(playerName, number);
+        super.addPlayer(playerName,option);
+    }
+
+    @Override
+    public void removePlayer(String playerName)
+    {
+        parsedPlayerOptions.remove(playerName);
+        super.removePlayer(playerName);
+    }
+
+    /**
+     * serves messages to players fetching them from the promise
+     * will repeat invocation for players that share the same option
+     */
+    @FunctionalInterface
+    public interface typeMessage<T> { BaseComponent [] get(T playerOption, Player player);}
+    public void log(typeMessage<T> messagePromise)
+    {
+        for (Map.Entry<String,T> en : parsedPlayerOptions.entrySet())
+        {
+            ServerPlayer player = playerFromName(en.getKey());
+            if (player != null)
+            {
+                BaseComponent [] messages = messagePromise.get(en.getValue(),player);
+                if (messages != null)
+                    sendPlayerMessage(player, messages);
+            }
+        }
+    }
+
+    /**
+     * guarantees that each message for each option will be evaluated once from the promise
+     * and served the same way to all other players subscribed to the same option
+     */
+    @FunctionalInterface
+    public interface typeMessageIgnorePlayer<T> { BaseComponent [] get(T playerOption);}
+    public void log(typeMessageIgnorePlayer<T> messagePromise)
+    {
+        Map<T, BaseComponent[]> cannedMessages = new HashMap<>();
+        for (Map.Entry<String,T> en : parsedPlayerOptions.entrySet())
+        {
+            ServerPlayer player = playerFromName(en.getKey());
+            if (player != null)
+            {
+                T option = en.getValue();
+                if (!cannedMessages.containsKey(option))
+                {
+                    cannedMessages.put(option,messagePromise.get(option));
+                }
+                BaseComponent [] messages = cannedMessages.get(option);
+                if (messages != null)
+                    sendPlayerMessage(player, messages);
+            }
+        }
+    }
+
+    public T parse(String value) // To be replaced by FromStringConverter in the settings rework
+    {
+        if (type == String.class)
+        {
+            return (T) value;
+        }
+        else if (type == boolean.class)
+        {
+            return (T) (Object) Boolean.parseBoolean(value);
+        }
+        else if (type == int.class)
+        {
+            return (T) (Object) Integer.parseInt(value);
+        }
+        else if (type == double.class)
+        {
+            return (T) (Object) Double.parseDouble(value);
+        }
+        else if (type.isEnum())
+        {
+            String ucValue = value.toUpperCase(Locale.ROOT);
+            return (T) Enum.valueOf((Class<? extends Enum>) type, ucValue);
+        }
+        else
+        {
+            return null;
+        }
```

Comment:
What if the converting function could be specified when constructing? For thought, though definitely not blocking this PR, can be done later.